### PR TITLE
[SPARK-14914] Fix Command too long for windows. Especially for test cases. 

### PR DESCRIPTION
**Pull request by taoli91**
## What changes were proposed in this pull request?

Windows has a limitation of 8192 characters on command line. This limitation will fail the test cases related to command builder in some test cases. This pull request adopt the method in Hadoop, which create a Jar that ref all the dependencies in the manifest to shorten the class path passing to the command line. 
## How was this patch tested?

Unit tests on windows and Linux. Note that this commit can't fix all the unit test errors on windows.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -1090,6 +1091,50 @@ private[spark] object Utils extends Logging {
     bytesToString(megabytes * 1024L * 1024L)
   }
 
+
+  /**
+   * Create a jar file at the given path, containing a manifest with a classpath
+   * that references all specified entries.
+   */
+  def createShortClassPath(tempDir: File, classPath: String) : String = {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Maybe, `def createShortClassPath(tempDir: File, classPath: String): String = {`

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 2]
```diff
@@ -1090,6 +1091,50 @@ private[spark] object Utils extends Logging {
...
+        .createJarWithClassPath(classPath, new Path(tempDir.getAbsolutePath), env)
+      val javaCpStr = javaCps(0) + javaCps(1)
+      logInfo("Shorten the class path to: " + javaCpStr)
+      javaCpStr
+    } else {
+      classPath
+    }
+  }
+
+  def createShortClassPath(classPath: String) : String = {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/util/Utils.scala]: Ditto.
```
def createShortClassPath(classPath: String): String = {
```

File: core/src/test/java/org/apache/spark/launcher/SparkLauncherSuite.java [hunk 3]
```diff
@@ -17,10 +17,13 @@
 
 package org.apache.spark.launcher;
 
+import java.io.File;
 import java.util.Arrays;
 import java.util.HashMap;
 import java.util.Map;
 
+import org.apache.commons.lang.SystemUtils;
+import org.apache.spark.util.Utils;
```
Comment (HyukjinKwon) [inline core/src/test/java/org/apache/spark/launcher/SparkLauncherSuite.java]: Maybe we should reorder imports. (See [databricks/scala-style-guide#imports](https://github.com/databricks/scala-style-guide#imports))

**Comment by vanzin** [comment]
I'm not a fan of this approach. This means that every invocation of the launcher code will create temp files and have to deal with cleaning it up; except the launcher cannot do it, because the jvm running SparkSubmit needs it. And now SparkSubmit needs to know to clean it up. That's adding a lot of complexity.

Setting the driver / executor classpaths in tests should be mostly unnecessary at this point, since the root pom file sets `SPARK_DIST_CLASSPATH` and that should be sufficient to propagate the needed jars to all child processes.

For the generic use case it is possible for someone could create a command line that exceeds the limit; but I think a different approach would be better. Maybe generating a temp config file, with the contents of both the user-provided config and the defaults from Spark's conf dir, but that has the same problem with "who deletes the temp conf file".

On a side note, there's a ton of style issues with your patch, please try to follow the style of the code you're modifying.

**Comment by vanzin** [comment]
Update: I thought about using argument files, but they're only supported by javac (`javac @argfile`) and not by java. So if fixing anything, I'm in favor of:

- removing the use of "driverJavaOptions" from unit tests since it shouldn't be needed - `SPARK_DIST_CLASSPATH` already handles it
- live with the fact that `SparkLauncher`'s generated command line  is limited to 8k characters on Windows

**Comment by srowen** [comment]
Let's close this PR

**Pull request closed without merging**